### PR TITLE
Added multi column partitioning for offline table

### PR DESCRIPTION
**Pull request by kmozaid**
## Description
This PR adds support for multi column partitioning for minion tasks and broker segment pruner.
## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] No

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] No

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] No (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)
## Release Notes
Segment can be partitioned on multiple columns when they are moved/created by minion tasks and when data is ingested via batch ingestion. Broker can also prune segments on multiple columns partition metadata.

### CODE REVIEW

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java [hunk 1]
```diff
@@ -54,18 +57,19 @@
  */
 public class PartitionSegmentPruner implements SegmentPruner {
   private static final Logger LOGGER = LoggerFactory.getLogger(PartitionSegmentPruner.class);
-  private static final PartitionInfo INVALID_PARTITION_INFO = new PartitionInfo(null, null);
+  private static final Map<String, PartitionInfo> INVALID_COLUMN_PARTITION_INFO =
+      Collections.singletonMap("null", new PartitionInfo(null, null));
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java:61]: (minor)
```suggestion
  private static final Map<String, PartitionInfo> INVALID_COLUMN_PARTITION_INFO_MAP = Collections.emptyMap();
```

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java [hunk 2]
```diff
@@ -83,20 +87,22 @@ public void init(IdealState idealState, ExternalView externalView, Set<String> o
     List<ZNRecord> znRecords = _propertyStore.get(segmentZKMetadataPaths, null, AccessOption.PERSISTENT, false);
     for (int i = 0; i < numSegments; i++) {
       String segment = segments.get(i);
-      PartitionInfo partitionInfo = extractPartitionInfoFromSegmentZKMetadataZNRecord(segment, znRecords.get(i));
-      if (partitionInfo != null) {
-        _partitionInfoMap.put(segment, partitionInfo);
+      Map<String, PartitionInfo> columnPartitionInfo =
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java:90]: (minor) rename it to `columnPartitionInfoMap`

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java [hunk 3]
```diff
@@ -83,20 +87,22 @@ public void init(IdealState idealState, ExternalView externalView, Set<String> o
...
   /**
    * NOTE: Returns {@code null} when the ZNRecord is missing (could be transient Helix issue). Returns
-   *       {@link #INVALID_PARTITION_INFO} when the segment does not have valid partition metadata in its ZK metadata,
-   *       in which case we won't retry later.
+   *       {@link #INVALID_COLUMN_PARTITION_INFO} when the segment does not have valid partition metadata in its ZK
+   *       metadata, in which case we won't retry later.
    */
   @Nullable
-  private PartitionInfo extractPartitionInfoFromSegmentZKMetadataZNRecord(String segment, @Nullable ZNRecord znRecord) {
+  private Map<String, PartitionInfo> extractPartitionInfoFromSegmentZKMetadataZNRecord(String segment,
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java:104]: (minor)
```suggestion
  private Map<String, PartitionInfo> extractColumnPartitionInfoMapFromSegmentZKMetadataZNRecord(String segment,
```

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java [hunk 4]
```diff
@@ -114,20 +120,25 @@ private PartitionInfo extractPartitionInfoFromSegmentZKMetadataZNRecord(String s
...
+      return INVALID_COLUMN_PARTITION_INFO;
     }
 
-    ColumnPartitionMetadata columnPartitionMetadata =
-        segmentPartitionMetadata.getColumnPartitionMap().get(_partitionColumn);
-    if (columnPartitionMetadata == null) {
-      LOGGER.warn("Failed to find column partition metadata for column: {}, segment: {}, table: {}", _partitionColumn,
-          segment, _tableNameWithType);
-      return INVALID_PARTITION_INFO;
+    Map<String, PartitionInfo> columnPartitionInfo = new HashMap<>();
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java:126]: Suggest returning a singleton map to accelerate the map lookup when there is only one entry in the map

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java [hunk 5]
```diff
@@ -165,9 +176,9 @@ public synchronized void refreshSegment(String segment) {
       }
       Set<String> selectedSegments = new HashSet<>();
       for (String segment : segments) {
-        PartitionInfo partitionInfo = _partitionInfoMap.get(segment);
-        if (partitionInfo == null || partitionInfo == INVALID_PARTITION_INFO || isPartitionMatch(filterExpression,
-            partitionInfo)) {
+        Map<String, PartitionInfo> columnPartitionInfo = _segmentColumnPartitionInfoMap.get(segment);
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java:179]: (minor) `columnPartitionInfoMap`

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java [hunk 6]
```diff
@@ -180,47 +191,57 @@ public synchronized void refreshSegment(String segment) {
...
             return true;
           }
         }
         return false;
       case EQUALS: {
         Identifier identifier = operands.get(0).getIdentifier();
-        if (identifier != null && identifier.getName().equals(_partitionColumn)) {
-          return partitionInfo._partitions.contains(
-              partitionInfo._partitionFunction.getPartition(operands.get(1).getLiteral().getFieldValue().toString()));
+        if (identifier != null && _partitionColumns.contains(identifier.getName())) {
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java:230]: Save one extra map lookup
```suggestion
        if (identifier != null) {
```

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java [hunk 7]
```diff
@@ -180,47 +191,57 @@ public synchronized void refreshSegment(String segment) {
...
+          return partitionInfo == null || partitionInfo._partitions.contains(
+              partitionInfo._partitionFunction.getPartition(operands.get(1).getLiteral().getFieldValue()));
         } else {
           return true;
         }
       }
       case IN: {
         Identifier identifier = operands.get(0).getIdentifier();
-        if (identifier != null && identifier.getName().equals(_partitionColumn)) {
+        if (identifier != null && _partitionColumns.contains(identifier.getName())) {
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java:240]:
```suggestion
        if (identifier != null) {
```

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java [hunk 8]
```diff
@@ -239,25 +260,29 @@ private boolean isPartitionMatch(Expression filterExpression, PartitionInfo part
...
-          if (isPartitionMatch(child, partitionInfo)) {
+          if (isPartitionMatch(child, columnPartitionInfo)) {
             return true;
           }
         }
         return false;
       case EQUALITY:
       case IN:
-        if (filterQueryTree.getColumn().equals(_partitionColumn)) {
+        if (_partitionColumns.contains(filterQueryTree.getColumn())) {
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java:281]: This check is redundant

**Review by Jackie-Jiang — COMMENTED** [review]
Shall we add a queries test (search for `*QueriesTest` for examples) or integration test to ensure it works end-to-end?

**Comment by kmozaid** [comment]
> Shall we add a queries test (search for `*QueriesTest` for examples) or integration test to ensure it works end-to-end?

I have added a test which prune segments based on query - 
https://github.com/apache/pinot/pull/8255/files#diff-9e8214b130b1a4a54157230e84674df2328018e7e7044a90a1f2a114ad56674bR353

and Sure let me see if I can add an integration test.

**Comment by kmozaid** [comment]
Hi @Jackie-Jiang Could you please check this PR if it can be merge now?

**Comment by Jackie-Jiang** [comment]
@kmozaid This is a great feature, and I want to ensure it can work end-to-end. With the `SegmentPrunerTest`, it ensures the pruner itself works, but multiple partitioning could fail in other places (I remember some classes explicitly check if there is only one partition column). I'd suggest adding an integration test to ensure it works end-to-end. You may refer to `NullHandlingIntegrationTest` on how to add an integration test with custom data.

**Comment by kmozaid** [comment]
Thanks @Jackie-Jiang , Sure, Let me add integration test.

**Comment by kmozaid** [comment]
> @kmozaid This is a great feature, and I want to ensure it can work end-to-end. With the `SegmentPrunerTest`, it ensures the pruner itself works, but multiple partitioning could fail in other places (I remember some classes explicitly check if there is only one partition column). I'd suggest adding an integration test to ensure it works end-to-end. You may refer to `NullHandlingIntegrationTest` on how to add an integration test with custom data.

Hi @Jackie-Jiang I have updated existing merge rollup minion task and realtime-to-offline minion task integration test which involves partitioning.Now these tests covered multi column partitioning flow. Please check if it is still required to write a separate integration tests.

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MergeRollupMinionClusterIntegrationTest.java [hunk 9]
```diff
@@ -97,7 +100,8 @@ public void setUp()
     TableConfig singleLevelConcatTableConfig =
         createOfflineTableConfig(SINGLE_LEVEL_CONCAT_TEST_TABLE, getSingleLevelConcatTaskConfig());
     TableConfig singleLevelRollupTableConfig =
-        createOfflineTableConfig(SINGLE_LEVEL_ROLLUP_TEST_TABLE, getSingleLevelRollupTaskConfig());
+        createOfflineTableConfig(SINGLE_LEVEL_ROLLUP_TEST_TABLE, getSingleLevelRollupTaskConfig(),
+            getMultiColumnSegmentPartitionConfig());
```
Comment (kmozaid) [inline pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MergeRollupMinionClusterIntegrationTest.java:104]: This `MergeRollupMinionClusterIntegrationTest` was started failing once I had added the multiple column segment partitioning config in table config. then I have fixed the `MergeRollupTaskGenerator` to make it run successfully.

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MergeRollupMinionClusterIntegrationTest.java [hunk 10]
```diff
@@ -178,6 +187,15 @@ private TableTaskConfig getMultiLevelConcatTaskConfig() {
     return new TableTaskConfig(Collections.singletonMap(MinionConstants.MergeRollupTask.TASK_TYPE, tableTaskConfigs));
   }
 
+  private SegmentPartitionConfig getMultiColumnSegmentPartitionConfig() {
+    Map<String, ColumnPartitionConfig> columnPartitionConfigMap = new HashMap<>();
+    ColumnPartitionConfig columnOneConfig = new ColumnPartitionConfig("murmur", 1);
+    columnPartitionConfigMap.put("AirlineID", columnOneConfig);
+    ColumnPartitionConfig columnTwoConfig = new ColumnPartitionConfig("murmur", 1);
+    columnPartitionConfigMap.put("Month", columnTwoConfig);
+    return new SegmentPartitionConfig(columnPartitionConfigMap);
```
Comment (kmozaid) [inline pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MergeRollupMinionClusterIntegrationTest.java:196]: Multi-columns partitioning config. This config is used by `MergeRollupTaskExecutor` (`SegmentProcessFramework`) to partition the data.

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/RealtimeToOfflineSegmentsMinionClusterIntegrationTest.java [hunk 11]
```diff
@@ -99,8 +103,16 @@ public void testRealtimeToOfflineSegmentsTask()
     List<SegmentZKMetadata> segmentsZKMetadata = _pinotHelixResourceManager.getSegmentsZKMetadata(_offlineTableName);
```
Comment (kmozaid) [inline pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/RealtimeToOfflineSegmentsMinionClusterIntegrationTest.java:103]: `testRealtimeToOfflineSegmentsTask` was very simple. It was creating one offline segments for each task. I have just added partitioning config to table and updated the assertions.

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/RealtimeToOfflineSegmentsMinionClusterIntegrationTest.java [hunk 12]
```diff
@@ -113,12 +125,21 @@ public void testRealtimeToOfflineSegmentsTask()
...
       segmentsZKMetadata = _pinotHelixResourceManager.getSegmentsZKMetadata(_offlineTableName);
-      numOfflineSegments++;
-      Assert.assertEquals(segmentsZKMetadata.size(), numOfflineSegments);
-      long expectedOfflineSegmentTimeMs = expectedWatermark - 86400000;
-      Assert.assertEquals(segmentsZKMetadata.get(i).getStartTimeMs(), expectedOfflineSegmentTimeMs);
-      Assert.assertEquals(segmentsZKMetadata.get(i).getEndTimeMs(), expectedOfflineSegmentTimeMs);
+      Assert.assertEquals(segmentsZKMetadata.size(), (numOfflineSegmentsPerTask * (i + 1)));
 
+      long expectedOfflineSegmentTimeMs = expectedWatermark - 86400000;
+      for (int j = (numOfflineSegmentsPerTask * i); j < segmentsZKMetadata.size(); j++) {
```
Comment (kmozaid) [inline pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/RealtimeToOfflineSegmentsMinionClusterIntegrationTest.java:131]: It checks only those segments which are generated by current task.

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/MergeTaskUtils.java [hunk 13]
```diff
@@ -61,9 +62,8 @@ public static TimeHandlerConfig getTimeHandlerConfig(TableConfig tableConfig, Sc
       return null;
     }
     DateTimeFieldSpec fieldSpec = schema.getSpecForTimeColumn(timeColumn);
-    Preconditions
-        .checkState(fieldSpec != null, "No valid spec found for time column: %s in schema for table: %s", timeColumn,
-            tableConfig.getTableName());
+    Preconditions.checkState(fieldSpec != null, "No valid spec found for time column: %s in schema for table: %s",
```
Comment (kmozaid) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/MergeTaskUtils.java:65]: Only formatting change.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/IntermediateSegment.java [hunk 14]
```diff
@@ -103,19 +100,6 @@ public IntermediateSegment(SegmentGeneratorConfig segmentGeneratorConfig) {
       }
     }
 
-    SegmentPartitionConfig segmentPartitionConfig = segmentGeneratorConfig.getSegmentPartitionConfig();
```
Comment (kmozaid) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/indexsegment/mutable/IntermediateSegment.java:106]: `IntermediateSegment` is not anywhere other than one test but I have still fixed considering some might use it  in future.

**Comment by kmozaid** [comment]
There are three other places where multicolumn partitioning config is not allowed.

1. `PinotLLCRealtimeSegmentManager.getPartitionMetadataFromTableConfig:710`
2. `LLRealtimeSegmentDataManager.setPartitionParameters:1425`
3. `PinotHadoopJobLauncher:HadoopSegmentPreprocessingJob.fetchPartitioningConfig:153`

This feature is NOT support by REALTIME tables so first 2 checks are still valid. The last one is specific to Hadoop pre processing job, I think we can skip for this feature as we don't really support `BoundedColumnValue` partitioner for hadoop;s job (https://github.com/apache/pinot/blob/master/pinot-plugins/pinot-batch-ingestion/v0_deprecated/pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/partitioners/PartitionFunctionFactory.java#L78). It is also deprecated.

**Comment by kmozaid** [comment]
Test Cases Covered -
1. Integration tests for Realtime to Offline Task with multi column partitioning
2. Integration tests for Merge Rollup task with multi column partitioning.
3. Unit tests for direct offline segment creation from file with multi column partitioning (merge rollup task integration tests setup does this).

**Comment by kmozaid** [comment]
Hi @Jackie-Jiang , Could you please provide review feedback? please also trigger the workflow again.

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java [hunk 15]
```diff
@@ -54,18 +57,18 @@
...
-  private static final PartitionInfo INVALID_PARTITION_INFO = new PartitionInfo(null, null);
+  private static final Map<String, PartitionInfo> INVALID_COLUMN_PARTITION_INFO_MAP = Collections.emptyMap();
 
   private final String _tableNameWithType;
-  private final String _partitionColumn;
+  private final Set<String> _partitionColumns;
   private final ZkHelixPropertyStore<ZNRecord> _propertyStore;
   private final String _segmentZKMetadataPathPrefix;
-  private final Map<String, PartitionInfo> _partitionInfoMap = new ConcurrentHashMap<>();
+  private final Map<String, Map<String, PartitionInfo>> _segmentColumnPartitionInfoMap = new ConcurrentHashMap<>();
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java:66]: Storing a map for each segment can potentially cause performance regression and extra memory footprint, and we already observe partition pruning becomes a bottleneck when a table contains a lot of segments. Let's add it as a separate implementation so that for single partition column case (probably 99% of the use case), we don't have to pay the extra cost. You may rename the current one to `SinglePartitionColumnSegmentPruner`, then add a `MultiPartitionColumnsSegmentPruner`

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java [hunk 16]
```diff
@@ -351,30 +350,39 @@ public List<PinotTaskConfig> generateTasks(List<TableConfig> tableConfigs) {
...
-                int partition = segmentPartitionMetadata.getPartitions(partitionColumn).iterator().next();
-                partitionToSegments.computeIfAbsent(partition, k -> new ArrayList<>()).add(selectedSegment);
+                String partitionId = StringUtils.join(partitionsBuffer, "_");
+                partitionToSegments.computeIfAbsent(partitionId, k -> new ArrayList<>()).add(selectedSegment);
               }
             }
 
-            for (Map.Entry<Integer, List<SegmentZKMetadata>> partitionToSegmentsEntry
-                : partitionToSegments.entrySet()) {
+            for (Map.Entry<String, List<SegmentZKMetadata>> partitionToSegmentsEntry : partitionToSegments.entrySet()) {
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java:385]: (minor) The keys are not used, use `valueSet()`

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java [hunk 17]
```diff
@@ -351,30 +350,39 @@ public List<PinotTaskConfig> generateTasks(List<TableConfig> tableConfigs) {
...
-          Preconditions.checkState(columnPartitionMap.size() == 1, "Cannot partition on multiple columns for table: %s",
-              tableConfig.getTableName());
-          Map.Entry<String, ColumnPartitionConfig> partitionEntry = columnPartitionMap.entrySet().iterator().next();
-          String partitionColumn = partitionEntry.getKey();
-
+          Set<String> partitionColumns = columnPartitionMap.keySet();
           for (List<SegmentZKMetadata> selectedSegmentsPerBucket : selectedSegmentsForAllBuckets) {
-            Map<Integer, List<SegmentZKMetadata>> partitionToSegments = new HashMap<>();
-            // Handle segments that have multiple partitions or no partition info
+            Map<String, List<SegmentZKMetadata>> partitionToSegments = new HashMap<>();
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java:361]: (minor) You may directly use `List<Integer>` as the key

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/RealtimeToOfflineSegmentsMinionClusterIntegrationTest.java (hunk 11 above)
Comment (Jackie-Jiang) [reply pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/RealtimeToOfflineSegmentsMinionClusterIntegrationTest.java:103]: I think the support is not added to realtime table, so how does it work on this test?

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java [hunk 18]
```diff
@@ -351,30 +350,39 @@ public List<PinotTaskConfig> generateTasks(List<TableConfig> tableConfigs) {
...
             List<SegmentZKMetadata> outlierSegments = new ArrayList<>();
             for (SegmentZKMetadata selectedSegment : selectedSegmentsPerBucket) {
               SegmentPartitionMetadata segmentPartitionMetadata = selectedSegment.getPartitionMetadata();
-              if (segmentPartitionMetadata == null
-                  || segmentPartitionMetadata.getPartitions(partitionColumn).size() != 1) {
+              List<Integer> partitionsBuffer = new ArrayList<>();
+              if (segmentPartitionMetadata != null && partitionColumns.equals(
+                  segmentPartitionMetadata.getColumnPartitionMap().keySet())) {
+                for (String partitionColumn : partitionColumns) {
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java:368]: We need to make `partitionColumns` a list to have consistent order

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/RealtimeToOfflineSegmentsMinionClusterIntegrationTest.java (hunk 11 above)
Comment (kmozaid) [reply pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/RealtimeToOfflineSegmentsMinionClusterIntegrationTest.java:103]: The `RealtimeToOfflineSegmentsTaskExecutor` uses the offline table config to create segments for offline table from realtime segments.

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java (hunk 15 above)
Comment (kmozaid) [reply pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java]: done.

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java (hunk 17 above)
Comment (kmozaid) [reply pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java]: done

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java (hunk 18 above)
Comment (kmozaid) [reply pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java:369]: done.

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java (hunk 16 above)
Comment (kmozaid) [reply pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java]: done.

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java [hunk 19]
```diff
@@ -18,266 +18,5 @@
...
-  private static class PartitionInfo {
-    final PartitionFunction _partitionFunction;
-    final Set<Integer> _partitions;
-
-    PartitionInfo(PartitionFunction partitionFunction, Set<Integer> partitions) {
-      _partitionFunction = partitionFunction;
-      _partitions = partitions;
-    }
-  }
+public interface PartitionSegmentPruner extends SegmentPruner {
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java:21]: Let's remove this file since this extra interface does not provide extra value (it should show as a rename from `PartitionSegmentPruner` to `SinglePartitionColumnSegmentPruner`)

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/SegmentPrunerFactory.java [hunk 20]
```diff
@@ -85,7 +87,7 @@ public static List<SegmentPruner> getSegmentPruners(TableConfig tableConfig,
             && LEGACY_PARTITION_AWARE_REALTIME_ROUTING.equalsIgnoreCase(routingTableBuilderName))) {
           PartitionSegmentPruner partitionSegmentPruner = getPartitionSegmentPruner(tableConfig, propertyStore);
           if (partitionSegmentPruner != null) {
-            segmentPruners.add(getPartitionSegmentPruner(tableConfig, propertyStore));
+            segmentPruners.add(partitionSegmentPruner);
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/SegmentPrunerFactory.java:90]: Good catch

File: pinot-integration-test-base/src/test/java/org/apache/pinot/integration/tests/BaseClusterIntegrationTest.java [hunk 21]
```diff
@@ -558,7 +563,8 @@ protected void waitForDocsLoaded(long timeoutMs, boolean raiseError) {
       @Override
       public Boolean apply(@Nullable Void aVoid) {
         try {
-          return getCurrentCountStarResult() == countStarResult;
+          long currentCountStarResult = getCurrentCountStarResult();
+          return currentCountStarResult == countStarResult;
```
Comment (Jackie-Jiang) [inline pinot-integration-test-base/src/test/java/org/apache/pinot/integration/tests/BaseClusterIntegrationTest.java:567]: (minor) seems unrelated

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MergeRollupMinionClusterIntegrationTest.java [hunk 22]
```diff
@@ -178,6 +187,15 @@ private TableTaskConfig getMultiLevelConcatTaskConfig() {
     return new TableTaskConfig(Collections.singletonMap(MinionConstants.MergeRollupTask.TASK_TYPE, tableTaskConfigs));
   }
 
+  private SegmentPartitionConfig getMultiColumnSegmentPartitionConfig() {
```
Comment (Jackie-Jiang) [inline pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MergeRollupMinionClusterIntegrationTest.java:190]: (minor)
```suggestion
  private SegmentPartitionConfig getMultiColumnsSegmentPartitionConfig() {
```

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java [hunk 23]
```diff
@@ -351,33 +352,41 @@ public List<PinotTaskConfig> generateTasks(List<TableConfig> tableConfigs) {
...
-          Preconditions.checkState(columnPartitionMap.size() == 1, "Cannot partition on multiple columns for table: %s",
-              tableConfig.getTableName());
-          Map.Entry<String, ColumnPartitionConfig> partitionEntry = columnPartitionMap.entrySet().iterator().next();
-          String partitionColumn = partitionEntry.getKey();
-
+          List<String> partitionColumns = new ArrayList<>(columnPartitionMap.keySet());
           for (List<SegmentZKMetadata> selectedSegmentsPerBucket : selectedSegmentsForAllBuckets) {
-            Map<Integer, List<SegmentZKMetadata>> partitionToSegments = new HashMap<>();
-            // Handle segments that have multiple partitions or no partition info
+            Map<List<Integer>, List<SegmentZKMetadata>> partitionToSegments = new LinkedHashMap<>();
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java:363]: I don't think we need a linked map here

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java [hunk 24]
```diff
@@ -351,33 +352,41 @@ public List<PinotTaskConfig> generateTasks(List<TableConfig> tableConfigs) {
...
             List<SegmentZKMetadata> outlierSegments = new ArrayList<>();
             for (SegmentZKMetadata selectedSegment : selectedSegmentsPerBucket) {
               SegmentPartitionMetadata segmentPartitionMetadata = selectedSegment.getPartitionMetadata();
-              if (segmentPartitionMetadata == null
-                  || segmentPartitionMetadata.getPartitions(partitionColumn).size() != 1) {
+              List<Integer> partitions = new ArrayList<>();
+              if (segmentPartitionMetadata != null && CollectionUtils.isEqualCollection(partitionColumns,
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java:368]: `CollectionUtils.isEqualCollection()` is much slower than `Set.equals()`. Suggest keeping both `partitionColumnSet` and `partitionColumnList`

**Review by Jackie-Jiang — APPROVED** [review]
LGTM with some minor comments. Thanks for adding the tests!

File: pinot-integration-test-base/src/test/java/org/apache/pinot/integration/tests/BaseClusterIntegrationTest.java (hunk 21 above)
Comment (kmozaid) [reply pinot-integration-test-base/src/test/java/org/apache/pinot/integration/tests/BaseClusterIntegrationTest.java:567]: yeah, I was debugging realtime cluster integration test and had to make this change for debug point. reverted it.

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java (hunk 23 above)
Comment (kmozaid) [reply pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java:363]: yes, we don't need linked map here, But the related test  https://github.com/apache/pinot/blob/master/pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/test/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGeneratorTest.java#L436 depends on the order of task to assert. But since we started using List<Integer> as key, the order got reversed and test was started failing. 

We can either use LinkedMap or fix test so that it is not depended on tasks order.

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java (hunk 24 above)
Comment (kmozaid) [reply pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java:368]: thanks.

File: pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java (hunk 23 above)
Comment (kmozaid) [reply pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java:363]: Fixed the test :)
